File: telegram-bot-game/minimax-m2/bot.js
const { Telegraf } = require('telegraf');

// Bot token - get from @BotFather
const BOT_TOKEN = process.env.BOT_TOKEN || 'YOUR_BOT_TOKEN_HERE';

const bot = new Telegraf(BOT_TOKEN);

// Game state storage - in-memory storage per chat
// Structure: { [chatId]: { targetNumber, minRange, maxRange, attempts, gameActive } }
const gameStates = new Map();

/**
 * Start a new game for a chat
 */
function startNewGame(ctx) {
  const chatId = ctx.chat.id;
  const state = gameStates.get(chatId) || {};

  // Use current range or default to 1-100
  const minRange = state.minRange || 1;
  const maxRange = state.maxRange || 100;

  // Generate random number
  const targetNumber = Math.floor(Math.random() * (maxRange - minRange + 1)) + minRange;

  // Update state
  const newState = {
    targetNumber,
    minRange,
    maxRange,
    attempts: 0,
    gameActive: true
  };

  gameStates.set(chatId, newState);

  ctx.reply(
    `🎮 New game started!\n` +
    `I'm thinking of a number between ${minRange} and ${maxRange}.\n` +
    `Try to guess it!`
  );
}

/**
 * Process a guess
 */
function processGuess(ctx, guess) {
  const chatId = ctx.chat.id;
  const state = gameStates.get(chatId);

  // Check if game is active
  if (!state || !state.gameActive) {
    ctx.reply('No active game. Use /newgame to start a new game!');
    return;
  }

  // Validate guess is within range
  if (guess < state.minRange || guess > state.maxRange) {
    ctx.reply(
      `Please guess a number between ${state.minRange} and ${state.maxRange}.`
    );
    return;
  }

  // Increment attempts
  state.attempts++;

  // Check if correct
  if (guess === state.targetNumber) {
    state.gameActive = false;
    gameStates.set(chatId, state);

    ctx.reply(
      `🎉 Correct! The number was ${state.targetNumber}.\n` +
      `You guessed it in ${state.attempts} ${state.attempts === 1 ? 'attempt' : 'attempts'}!\n` +
      `Type /newgame to play again or /range to change the range.`
    );
  } else if (guess < state.targetNumber) {
    ctx.reply('📈 Higher!');
  } else {
    ctx.reply('📉 Lower!');
  }
}

// Command: /start
bot.start((ctx) => {
  ctx.reply(
    'Welcome to the Number Guessing Game! 🎮\n' +
    '\n' +
    'I\'ll think of a number and you try to guess it.\n' +
    'I\'ll tell you if your guess is higher or lower.\n' +
    '\n' +
    'Commands:\n' +
    '/newgame - Start a new game\n' +
    '/range <min> <max> - Set the number range (e.g., /range 1 1000)\n' +
    '/help - Show this help message\n' +
    '\n' +
    'Let\'s start! Type /newgame to begin.'
  );
});

// Command: /help
bot.help((ctx) => {
  ctx.reply(
    '📖 How to Play:\n' +
    '\n' +
    '1. Start a game with /newgame\n' +
    '2. I\'ll pick a random number in the current range\n' +
    '3. Guess the number and I\'ll say "higher" or "lower"\n' +
    '4. Try to guess in as few attempts as possible!\n' +
    '\n' +
    '🔧 Commands:\n' +
    '/newgame - Start a new game\n' +
    '/range <min> <max> - Change the number range\n' +
    '/help - Show this help\n' +
    '\n' +
    'Current range is 1-100. Change it with /range!'
  );
});

// Command: /newgame
bot.command('newgame', (ctx) => {
  startNewGame(ctx);
});

// Command: /range
bot.command('range', (ctx) => {
  const args = ctx.message.text.split(' ').slice(1);

  if (args.length !== 2) {
    const state = gameStates.get(ctx.chat.id);
    const currentMin = state?.minRange || 1;
    const currentMax = state?.maxRange || 100;
    ctx.reply(
      `Current range: ${currentMin} - ${currentMax}\n` +
      `Usage: /range <min> <max>\n` +
      `Example: /range 1 1000`
    );
    return;
  }

  const min = parseInt(args[0]);
  const max = parseInt(args[1]);

  // Validate input
  if (isNaN(min) || isNaN(max)) {
    ctx.reply('Please provide valid numbers. Example: /range 1 100');
    return;
  }

  if (min >= max) {
    ctx.reply('The minimum must be less than the maximum. Example: /range 1 100');
    return;
  }

  if (min < -1000000 || max > 1000000) {
    ctx.reply('Please keep numbers between -1,000,000 and 1,000,000.');
    return;
  }

  // Update or create state with new range
  const state = gameStates.get(ctx.chat.id) || {};
  state.minRange = min;
  state.maxRange = max;
  state.gameActive = false; // Deactivate current game if any
  gameStates.set(ctx.chat.id, state);

  ctx.reply(
    `✅ Range updated to ${min} - ${max}\n` +
    `Type /newgame to start a new game with this range!`
  );
});

// Handle regular messages (guesses)
bot.on('text', (ctx) => {
  const message = ctx.message.text;

  // Skip if it's a command
  if (message.startsWith('/')) {
    return;
  }

  // Try to parse as number
  const guess = parseInt(message);

  // Check if it's a valid number
  if (isNaN(guess)) {
    ctx.reply('Please send a valid number to guess.');
    return;
  }

  processGuess(ctx, guess);
});

// Error handling
bot.catch((err, ctx) => {
  console.log(`Error for ${ctx.updateType}:`, err);
  ctx.reply('Sorry, something went wrong! Please try again.');
});

// Launch bot
bot.launch().then(() => {
  console.log('🤖 Bot is running...');
});

// Enable graceful stop
process.once('SIGINT', () => bot.stop('SIGINT'));
process.once('SIGTERM', () => bot.stop('SIGTERM'));

File: telegram-bot-game/gpt-5-codex-high/bot.py
"""Telegram number guessing bot."""

from __future__ import annotations

import logging
import os
import random
import sys
from dataclasses import dataclass
from typing import Dict, Tuple

from telegram import Update
from telegram.ext import (
    Application,
    CommandHandler,
    ContextTypes,
    MessageHandler,
    filters,
)


logger = logging.getLogger(__name__)


@dataclass
class GameState:
    """In-memory state for a single chat/user pair."""

    lower: int = 1
    upper: int = 100
    target: int = 0
    attempts: int = 0
    active: bool = False


def _state_key(chat_id: int | None, user_id: int | None) -> Tuple[int, int]:
    if chat_id is None or user_id is None:
        raise ValueError("Both chat_id and user_id must be available")
    return chat_id, user_id


def get_game_state(update: Update, context: ContextTypes.DEFAULT_TYPE) -> GameState:
    key = _state_key(update.effective_chat.id, update.effective_user.id)
    store: Dict[Tuple[int, int], GameState] = context.application.bot_data.setdefault(
        "game_states", {}
    )
    if key not in store:
        store[key] = GameState()
    return store[key]


def start_new_round(state: GameState, lower: int | None = None, upper: int | None = None) -> None:
    if lower is not None and upper is not None:
        state.lower, state.upper = lower, upper
    state.target = random.randint(state.lower, state.upper)
    state.attempts = 0
    state.active = True
    logger.debug("Started new round with target=%s", state.target)


def format_range(state: GameState) -> str:
    return f"between {state.lower} and {state.upper}"


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    state = get_game_state(update, context)
    start_new_round(state)
    await update.message.reply_text(
        "Welcome to the number guessing game!\n"
        f"I'm thinking of a number {format_range(state)}."
        "\nSend guesses as plain numbers, or use /range to change the limits."
    )


async def new_game(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    state = get_game_state(update, context)
    start_new_round(state)
    await update.message.reply_text(
        f"New game started! Guess the number {format_range(state)}."
    )


async def set_range(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if len(context.args) != 2:
        await update.message.reply_text("Usage: /range <min> <max>")
        return
    try:
        lower = int(context.args[0])
        upper = int(context.args[1])
    except ValueError:
        await update.message.reply_text("Please supply two whole numbers, e.g. /range 1 500")
        return
    if lower >= upper:
        await update.message.reply_text("The first number must be smaller than the second.")
        return
    state = get_game_state(update, context)
    start_new_round(state, lower, upper)
    await update.message.reply_text(
        f"Range updated to {lower}-{upper}. I'm thinking of a new number!"
    )


async def help_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    await update.message.reply_text(
        "Commands:\n"
        "/start - start playing\n"
        "/newgame - restart with the current range\n"
        "/range <min> <max> - set a new range and restart\n"
        "/help - show this message\n\n"
        "Send a number at any time to make a guess."
    )


async def handle_guess(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if not update.message or not update.message.text:
        return
    state = get_game_state(update, context)
    text = update.message.text.strip()
    try:
        guess = int(text)
    except ValueError:
        await update.message.reply_text("I didn't understand that number. Try again or use /help.")
        return

    if not state.active:
        await update.message.reply_text(
            "You're not currently in a game. Use /newgame to start again."
        )
        return

    if guess < state.lower or guess > state.upper:
        await update.message.reply_text(
            f"Please guess a number {format_range(state)}."
        )
        return

    state.attempts += 1

    if guess < state.target:
        await update.message.reply_text("Higher!")
    elif guess > state.target:
        await update.message.reply_text("Lower!")
    else:
        state.active = False
        await update.message.reply_text(
            f"Correct! The number was {state.target}. "
            f"You needed {state.attempts} attempts."
            "\nUse /newgame to play again or /range to change the limits."
        )


def main() -> None:
    logging.basicConfig(
        format="%(asctime)s %(levelname)s %(name)s: %(message)s",
        level=logging.INFO,
    )

    token = os.environ.get("TELEGRAM_BOT_TOKEN")
    if not token:
        logger.error("Missing TELEGRAM_BOT_TOKEN environment variable")
        sys.exit(1)

    application = Application.builder().token(token).build()

    application.add_handler(CommandHandler("start", start))
    application.add_handler(CommandHandler("newgame", new_game))
    application.add_handler(CommandHandler("range", set_range))
    application.add_handler(CommandHandler("help", help_command))
    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_guess))

    logger.info("Bot started")
    application.run_polling()


if __name__ == "__main__":
    main()
